analyzers/src/SonarAnalyzer.ShimLayer/LightupHelpers.cs

```diff
@@ -1,874 +1,890 @@
-            throw new InvalidOperationException();
-        }
+            if (!typeof(TProperty).GetTypeInfo().IsAssignableFrom(property.PropertyType.GetTypeInfo()))
+            {
+                throw new InvalidOperationException();
+            }
 
-        var property = type.GetTypeInfo().GetDeclaredProperty(propertyName);
-        if (property == null)
-        {
-            return FallbackAccessor;
+            Expression<Func<TProperty>> expression =
+                Expression.Lambda<Func<TProperty>>(
+                    Expression.Call(null, property.GetMethod));
+            return expression.Compile();
         }
-        var syntaxParameter = Expression.Parameter(typeof(TSyntax), "syntax");
-        Expression instance =
-            type.GetTypeInfo().IsAssignableFrom(typeof(TSyntax).GetTypeInfo())
-            ? (Expression)syntaxParameter
-            : Expression.Convert(syntaxParameter, type);
 
-        Expression body = Expression.Call(instance, property.GetMethod);
-
-        if (!typeof(TProperty).GetTypeInfo().IsAssignableFrom(property.PropertyType.GetTypeInfo()))
+        public static Func<TSyntax, TProperty> CreateSyntaxPropertyAccessor<TSyntax, TProperty>(Type type, string propertyName)
         {
-            body = Expression.Convert(body, typeof(TProperty));
-        }
+            TProperty FallbackAccessor(TSyntax syntax)
+            {
+                if (syntax == null)
+                {
+                    // Unlike an extension method which would throw ArgumentNullException here, the light-up
+                    // behavior needs to match behavior of the underlying property.
+                    throw new NullReferenceException();
+                }
+
+                return default;
+            }
 
-        return Expression.Lambda<Func<TSyntax, TProperty>>(body, syntaxParameter).Compile();
-    }
+            if (type == null)
+            {
+                return FallbackAccessor;
+            }
 
-    internal static Func<TSyntax, TArg, TProperty> CreateSyntaxPropertyAccessor<TSyntax, TArg, TProperty>(Type type, Type argumentType, string accessorMethodName)
-    {
-        static TProperty FallbackAccessor(TSyntax syntax, TArg argument)
-        {
-            if (syntax == null)
+            if (!typeof(TSyntax).GetTypeInfo().IsAssignableFrom(type.GetTypeInfo()))
             {
-                // Unlike an extension method which would throw ArgumentNullException here, the light-up
-                // behavior needs to match behavior of the underlying property.
-                throw new NullReferenceException();
+                throw new InvalidOperationException();
             }
 
-            return default;
-        }
+            var property = type.GetTypeInfo().GetDeclaredProperty(propertyName);
+            if (property == null)
+            {
+                return FallbackAccessor;
+            }
+            var syntaxParameter = Expression.Parameter(typeof(TSyntax), "syntax");
```

Comment:
I think // Sonar begin should start at line 324. From what I can see, this is were the update code starts.